File: backend-typescript/src/todo/services/todo.service.ts
import { Injectable } from '@nestjs/common';
import { TodoRepository } from '../repositories/todo.repository';
import { CreateTodoDto } from '../dtos/create-todo.dto';
import { UpdateTodoDto } from '../dtos/update-todo.dto';
import { Todo } from '../entities/todo.entity';

@Injectable()
export class TodoService {
  constructor(private readonly todoRepository: TodoRepository) {}

  async findAll(): Promise<Todo[]> {
    return this.todoRepository.findAll();
  }

  async findById(id: number): Promise<Todo | null> {
    return this.todoRepository.findById(id);
  }

  async create(createTodoDto: CreateTodoDto): Promise<Todo> {
    return this.todoRepository.create(createTodoDto);
  }

  async update(id: number, updateTodoDto: UpdateTodoDto): Promise<Todo | null> {
    return this.todoRepository.update(id, updateTodoDto);
  }

  async delete(id: number): Promise<boolean> {
    return this.todoRepository.delete(id);
  }
} 

File: backend-python/src/services/todo_service.py
from typing import List, Optional
from src.models.todo import Todo
from src.repositories.todo_repository import TodoRepository
from src.schemas import TodoRequest


class TodoService:
    """Service layer for todo business logic."""

    def __init__(self, repository: TodoRepository):
        self.repository = repository

    def get_all_todos(self) -> List[Todo]:
        """Get all todos."""
        return self.repository.find_all()

    def get_todo_by_id(self, todo_id: int) -> Optional[Todo]:
        """Get a todo by ID."""
        return self.repository.find_by_id(todo_id)

    def create_todo(self, todo_request: TodoRequest) -> Todo:
        """Create a new todo."""
        todo = Todo(
            task=todo_request.task,
            completed=todo_request.completed if todo_request.completed is not None else False
        )
        return self.repository.save(todo)

    def update_todo(self, todo_id: int, todo_request: TodoRequest) -> Optional[Todo]:
        """Update an existing todo."""
        existing_todo = self.repository.find_by_id(todo_id)
        if existing_todo is None:
            return None

        existing_todo.task = todo_request.task
        existing_todo.completed = todo_request.completed if todo_request.completed is not None else False

        return self.repository.save(existing_todo)

    def delete_todo(self, todo_id: int) -> bool:
        """Delete a todo by ID."""
        todo = self.repository.find_by_id(todo_id)
        if todo is None:
            return False

        self.repository.delete(todo)
        return True
